phper/src/types.rs

```diff
@@ -10,162 +10,282 @@
+    pub const DOUBLE: TypeInfo = TypeInfo::from_type_mask(IS_DOUBLE);
+    pub const FALSE: TypeInfo = TypeInfo::from_type_mask(IS_FALSE);
+    pub const ITERABLE: TypeInfo = TypeInfo::from_type_mask(IS_ITERABLE);
+    pub const LONG: TypeInfo = TypeInfo::from_type_mask(IS_LONG);
+    pub const MIXED: TypeInfo = TypeInfo::from_type_mask(IS_MIXED);
+    pub const NEVER: TypeInfo = TypeInfo::from_type_mask(IS_NEVER);
+    pub const NULL: TypeInfo = TypeInfo::from_type_mask(IS_NULL);
+    pub const OBJECT: TypeInfo = TypeInfo::from_type_mask(IS_OBJECT);
+    pub const RESOURCE: TypeInfo = TypeInfo::from_type_mask(IS_RESOURCE);
+    pub const STRING: TypeInfo = TypeInfo::from_type_mask(IS_STRING);
+    pub const TRUE: TypeInfo = TypeInfo::from_type_mask(IS_TRUE);
+    pub const UNDEF: TypeInfo = TypeInfo::from_type_mask(IS_UNDEF);
+    pub const VOID: TypeInfo = TypeInfo::from_type_mask(IS_VOID);
 }
 
 impl TypeInfo {
-    pub const fn from_raw(t: u32) -> Self {
-        Self { t }
+    pub fn varadic(&mut self) -> &Self {
+        self.inner.type_mask |= _ZEND_IS_VARIADIC_BIT;
+
+        self
+    }
+
+    pub const fn from_type_mask(mask: u32) -> Self {
+        Self {
+            inner: ZendType {
+                ptr: std::ptr::null::<std::ffi::c_void>() as *mut std::ffi::c_void,
+                type_mask: mask,
+            },
+        }
     }
 
-    pub const fn into_raw(self) -> u32 {
-        self.t
+    pub fn from_zval(zval: &ZVal) -> Self {
+        let mask = unsafe { phper_z_type_info_p(zval.as_ptr()) };
+        if (mask & IS_OBJECT) != 0 {
+            let name_ptr = unsafe {
+                let obj = phper_z_obj_p(zval.as_ptr());
+                let ce = (*obj).ce;
+                phper_zend_object_release(obj);
+                (*ce).name as *mut std::ffi::c_void
+            };
+
+            Self {
+                inner: ZendType {
+                    ptr: name_ptr,
+                    type_mask: mask | _ZEND_TYPE_NAME_BIT,
+                },
+            }
+        } else {
+            Self::from_type_mask(mask)
+        }
+    }
+
+    pub const fn from_class_entry(ce: &zend_class_entry) -> Self {
+        let name_ptr = ce.name as *mut std::ffi::c_void;
+
+        Self {
+            inner: ZendType {
+                ptr: name_ptr,
+                type_mask: _ZEND_TYPE_NAME_BIT,
+            },
+        }
+    }
+
+    pub const fn from_raw(inner: ZendType) -> Self {
+        Self { inner }
+    }
+
+    pub const fn into_raw(self) -> ZendType {
+        self.inner
     }
 
     pub const fn undef() -> TypeInfo {
-        Self::from_raw(IS_UNDEF)
+        Self::from_type_mask(IS_UNDEF)
     }
 
     pub const fn null() -> TypeInfo {
-        Self::from_raw(IS_NULL)
+        Self::from_type_mask(IS_NULL)
     }
 
     pub const fn bool(b: bool) -> TypeInfo {
-        Self::from_raw(if b { IS_TRUE } else { IS_FALSE })
+        Self::from_type_mask(if b { IS_TRUE } else { IS_FALSE })
     }
```

Comment:
should we rename this to `from_bool`? it doesn't make sense that it needs a value.